fabtests/common/hmem_cuda.c

```diff
@@ -49,15 +50,55 @@ struct cuda_ops {
 	cudaError_t (*cudaMemset)(void *ptr, int value, size_t count);
 	const char *(*cudaGetErrorName)(cudaError_t error);
 	const char *(*cudaGetErrorString)(cudaError_t error);
+	CUresult (*cuPointerSetAttribute)(void *data,
+					  CUpointer_attribute attribute,
+					  CUdeviceptr ptr);
+	CUresult (*cuGetErrorName)(CUresult error, const char** pStr);
+	CUresult (*cuGetErrorString)(CUresult error, const char** pStr);
 };
 
 static struct cuda_ops cuda_ops;
 static void *cudart_handle;
+static void *cuda_handle;
+
+/**
+ * Since function names can get redefined in cuda.h/cuda_runtime.h files,
+ * we need to do this stringifying to get the latest function name from
+ * the header files.  For example, cuda.h may have something like this:
+ * #define cuMemFree cuMemFree_v2
+ * We want to make sure we find cuMemFree_v2, not cuMemFree.
+ */
+#define STRINGIFY2(x) #x
+#define STRINGIFY(x)  STRINGIFY2(x)
 
 #define CUDA_ERR(err, fmt, ...) \
 	FT_ERR(fmt ": %s %s", ##__VA_ARGS__, cuda_ops.cudaGetErrorName(err), \
 	       cuda_ops.cudaGetErrorString(err))
 
+static void ft_cuda_driver_api_print_error(CUresult cu_result, char *cuda_api_name)
+{
+	const char *cu_error_name;
+	const char *cu_error_str;
+	cuda_ops.cuGetErrorName(cu_result, &cu_error_name);
+	cuda_ops.cuGetErrorString(cu_result, &cu_error_str);
+	FT_ERR("%s failed: %s:%s\n",
+		   cuda_api_name, cu_error_name, cu_error_str);
+}
+
+static int ft_cuda_pointer_set_attribute(void *buf)
+{
+	int true_flag = 1;
+	CUresult cu_result;
+	cu_result = cuda_ops.cuPointerSetAttribute((void *) &true_flag,
```

Comment:
Didn't you say this was insufficient anyway? So device synchronization is being done instead?

Reply:
This is still required to ensure the synchronous data movement between devices. Because we only do synchronize in libfabric if the buffer is not allocated/registered. For cudaMemcpy between device buffers, there is no cudaDeviceSynchronize inside libfabric.